java/arcs/core/crdt/CrdtSet.kt

```diff
@@ -78,10 +88,15 @@ class CrdtSet<T : Referencable>(
         }
 
         _data.values.forEach { (id, myEntry) ->
-            if (id !in other.values && other.versionMap isDominatedBy myEntry.versionMap) {
-                // Value was added by this model.
-                mergedData.values[id] = myEntry
-                fastForwardOp.added += myEntry
+            if (id !in other.values) {
+                if (other.versionMap isDominatedBy myEntry.versionMap) {
+                    // Value was added by this model.
+                    mergedData.values[id] = myEntry
+                    fastForwardOp.added += myEntry
+                } else {
+                    // Value was removed by other model.
+                    fastForwardOp.removed += myEntry.value
+                }
```

Comment:
This fixes the issue where removals due to a merge were not reported

Reply:
This seems right to me, but I'd wait for @csilvestrini to chime in, as he's the inventor of the FastForwardOp stuff.

Reply:
I think this is wrong, and the old code was wrong too. Here's the equivalent TypeScript code for comparison:

```
    for (const [id, thisEntry] of Object.entries(this.model.values)) {
      if (!other.values[id] && !dominates(other.version, thisEntry.version)) {
        // Value was added by this model.
        merged[id] = thisEntry;
        fastForwardOp.added.push([thisEntry.value, thisEntry.version]);
      }
    }
```

Essentially we're looking for values only present in the model for `this` that are missing from `other`, and trying to work out whether they were added by `this` or removed by `other`. The removal will "win" over the addition if its version map dominates the version map of the addition.

So if `myEntry.versionMap isDominatedBy other.versionMap` then `other` saw the addition op, and subsequently removed it. In that case, we don't add it to `mergedData`, and don't need to tell `other` to do anything about it with a fast forward op.

If it is _not_ dominated, then we add it to `mergedData` and add it to the fast forward op. 

Not that `a isDominatedBy b` is not equivalent to `b !isDominatedBy a`.

I think the correct fix here is to revert this change and swap the args in the if-statement like so:

```
if (id !in other.values && myEntry.versionMap !isDominatedBy other.versionMap) {
```

Reply:
This change doesn't address the issue that I was trying to resolve here: there was nothing in the logic that added a removal operation to the merge changes when an item was *removed* from the local set due to its removal in the merged-in data.

The actual merged *data* correctly reflected this change, but the operations that were returned from the merge operation did not include it. 

This means that the following didn't work:

```
handleAProcessA.store(entity)
```

```
processBHandle.fetchAll() // contains entity
processBHandle.remove(entity)
processBHandle.fetchAll() // does not contain entity
```

```
processAHandle.fetchAll() // should not contain the entity, but it still does
```

Reply:
Put another way: This PR was not trying to change anything about how the `merge` was applied, only how it was reported to listeners.

Reply:
As discussed in chat, this is the fix I am proposing. I verified it fixes the DifferentHandleManagerDifferentStoresTest

```
_data.values.forEach { (id, myEntry) ->
            if (id !in other.values && other.versionMap isDominatedBy myEntry.versionMap) {
                // Value was added by this model.
                mergedData.values[id] = myEntry
                fastForwardOp.added += myEntry
            }
        }

        val otherOperations = if (
            fastForwardOp.added.isNotEmpty() ||
            fastForwardOp.removed.isNotEmpty() ||
            oldClock isDominatedBy newClock
        ) {
            Operations<Data<T>, IOperation<T>>(fastForwardOp.simplify().toMutableList())
        } else {
            Operations<Data<T>, IOperation<T>>(mutableListOf())
        }

        val myChange = if (mergedData.equals(this._data)) {
            Operations<Data<T>, IOperation<T>>(mutableListOf())
        } else {
            CrdtChange.Data<Data<T>, IOperation<T>>(mergedData)
        }

        this._data = mergedData

        return MergeChanges(
            modelChange = myChange, otherChange = otherOperations
        )
```

Reply:
It looks like the implementation of `isDominatedBy` is incorrect (but that bug was causing this code to work correctly), so what I was saying was kinda true and kinda not. 

I've put together a PR to fix up isDominatedBy: https://github.com/PolymerLabs/arcs/pull/5102